### PR TITLE
feat/observers: update to new ui

### LINKED ISSUE #225: [Observers] Add / edit / delete observer - update to new UI
**Depends on** https://github.com/code4romania/monitorizare-vot-ong/issues/224

UI is available here (make it as similar as possible to the new one): https://www.figma.com/file/61NHR7fwmV1Dsv4Pb6F8T7/code4ro_mv_user-management_2018.04.21-(Copy)?node-id=754%3A0

- [ ] new button for adding an observer 
- [ ] add a new page for adding / editing observer details
- [ ] add the available actions button&dropdown in the observer list - available actions are edit, delete, reset password and send notification

Please use bootstrap classes as much as possible and avoid adding custom css styles to keep the codebase clean & consistent. The new UI should be as similar as possible to the one in figma, not pixel perfect.

Also, make sure all texts are extracted in translation files.

### PR DESCRIPTION
**Pull request by Andrei0872**
closes #225
closes https://github.com/code4romania/monitorizare-vot-ong/issues/224

---

 Since this is not a very small PR, I will go through the relevant additions. Although it went pretty well design-wise, there are a few things left unfinished as far as the functionality is concerned. 

### `base button`

* it is a generic component that encapsulates the styles that are encountered across the app(this resulted from what I've noticed in the designs)
* it supports variants and specific customizations, depending on the requirements

### `table-container`

* it includes `container` in its name because, apart from the table itself, it also container 2 other zones - the one where the filters are added and the pagination; this was the current case, but these zones can be anything, as long as they're provided accordingly(using `header-area`, `footer-area` attributes)
* it includes the logic related to checkbox selection
* it provides a way for the consumer to provide the cell's content based on certain conditions

https://github.com/code4romania/monitorizare-vot-ong/blob/e00d39e2e7e2bae31c3bb086ab1e29e9a96b79b6/src/app/table/table-container/table-container.component.html#L30-L36

```html
<ng-template let-r="row" let-c="column" >
 <!-- content can be anything, it usually involves some logic around `r` and `c` --> 
 {{ r[c] }}
<ng-template>
```

 ### `base-checkbox`

* it encapsulates the logic and styles such that it can be used in multiple places
* if needed, it can support variants, just as the `base-button` component does

### `base-dropdown`

* this is the widget from the _Actions_ column
* it can be used in multiple places across the app; it receives a configuration array and based on it, the component gives the consumer the choice to choose the content of each button 

---

## Unfinished things/questions

1. It is required to support smaller screens? if yes, how should the page look?

2. What's the API for the _sort by a cell_ feature? I noticed the `Last Login` should have this feature.

3. Where should the loading icon be placed? Should it be handled inside `table-container`?

4. What's the API for the `Import`/`Export` buttons?

5. What's the API for sending notifications/deleting multiple observers? For example, when multiple rows are selected, the user can send notification to each of them or delete them all.

6. Referring to the `observatori/new` and `observatori/edit/:id` views, I think there is a small incompatibility between the design the and API's response. For example, when editing an observer, the API response might look like

```json
{
id: 2
lastSeen: null
name: "irina"
ngo: "Code4Romania"
numberOfNotes: 0
numberOfPollingStations: 0
phone: "0700000000"
}
```

and in the form we have `firstName`, `lastName`, `phoneNumber`.

7. This is derived from `6`, I don't know what would the API be when a form is submitted in `/add` / `/edit` views.

### CODE REVIEW

**Comment by aniri** [comment]
woooh @Andrei0872 thanks for the PR. you covered at least two issues in it :smiley: 

to answer your questions:
> It is required to support smaller screens? if yes, how should the page look?

It's just a nice-to-have and we do not have a design for this. If you have something in mind, you can try it :D

> What's the API for the sort by a cell feature? I noticed the Last Login should have this feature.

The api does not support this yet, so we can leave it as a future improvement.

> Where should the loading icon be placed? Should it be handled inside table-container?

Wherever it is easier to be placed, as it was not covered in the design.

> What's the API for the Import/Export buttons?

the export functionality is not available yet. but the endpoint for import is `POST ​/api​/v1​/observer​/import`. it needs an ngoId and a file. so pressing that button should open a new page (not covered in design but should be similar to the one for adding a new observer) with the input field for ngo id and file uploader. you can see this in the old version of the screen if you press 'add observer' .

> What's the API for sending notifications/deleting multiple observers? For example, when multiple rows are selected, the user can send notification to each of them or delete them all.

This is not ready yet, and we have a separate UI issues for implementing this https://github.com/code4romania/monitorizare-vot-ong/issues/227 . We can leave them as they are for now :)

> Referring to the observatori/new and observatori/edit/:id views, I think there is a small incompatibility between the design the and API's response. For example, when editing an observer, the API response might look like
> 
> {
> id: 2
> lastSeen: null
> name: "irina"
> ngo: "Code4Romania"
> numberOfNotes: 0
> numberOfPollingStations: 0
> phone: "0700000000"
> }
> and in the form we have firstName, lastName, phoneNumber.
> 
> This is derived from 6, I don't know what would the API be when a form is submitted in /add / /edit views.

Yes, there are some incompatibilities between the design and the data model from the api:
- the first name + last name inputs should be replaced with just one `name` , as we currently do not support two :(
- the password field should not be hidden
- the add / edit observer form should have `name`, `phone number` and `password` (not hidden), all fields mandatory and call the `POST ​/api​/v1​/observer` for add and `PUT ​/api​/v1​/observer` for edit 
- the reset password option should be added in the actions dropdown for each observer (but not as a bulk action) and work similar to the old implementation (it is not covered in the design :( ) with a popup (similar to what we have in the [styleguide](https://www.figma.com/file/61NHR7fwmV1Dsv4Pb6F8T7/code4ro_mv_user-management_2018.04.21-(Copy)?node-id=816%3A217) but with an input field as in the old implementation) and uses the `POST ​/api​/v1​/observer​/reset` endpoint

Other things I noticed while checking this out and wuld be nice-to-haves:
- could we also add the `numberOfPollingStations` as a field in the table called `Visited polling stations`?
- could you extract all texts in the translation files?
- please leave the font as it was, we'll probably update it later :smiley: 

Thank you!!

**Comment by Andrei0872** [comment]
@aniri thanks for the feedback! It is okay if I don't deal with API calls and only provide the handler where that a call would take place? I'm suggesting this since I'm not sure I'll have all the necessary permissions for each action.

Without the external API-related tasks, this would be my _todo_ list:

* [ ]  numberOfPollingStations as a field in the table called Visited polling stations
* [ ] remove custom font
* [ ] add translations: apart from `numberOfPollingStations`, every user-facing text should already have its translation in the files
* [ ] _observers_  page should support smaller screens - challenge accepted :sunglasses:
* [ ] add only the design for _sort-by-a-cell_ feature
* [ ] add loading icon inside the table
* [ ] replace `firstName` and `lastName` fields with `name` (which is `firstName` + `lastName`)
* [ ] in both `/add` and `edit` views, there should be: `name`, `phone number` and `password`(not hidden), all required; 
* [ ] add _reset password_ option

Regarding to the _reset password_ option, I'm not sure I fully understood how it should look like. The way I understood it: it should be another button in the dropdown list and when clicked, there should be an alert box which should contain an input field where the new password would be inserted. Is it accurate?

Thank you!

**Comment by aniri** [comment]
@Andrei0872 yep, that is accurate! :rocket: Thanks!

You are also correct about the reset password feature :) It was implemented before so you can also check it out.

It would be great if you could add the add/edit endpoints based on swagger. Even if not tested. It will be easier to test and adjust them if everything is in place :)

**Comment by Andrei0872** [comment]
@aniri sure, where can I find the Swagger docs for the current API?

**Comment by aniri** [comment]
@Andrei0872 

here: https://api.votemonitor.org/swagger/index.html

and the test environment with the old web app: https://monitorizare-vot-ong-git-develop.code4romania.vercel.app so you can check thr old observers page and related requests ;)

**Comment by Andrei0872** [comment]
Hello @aniri , I've completed all the _todos_ above. 

A few things worth mentioning:

* Calling the APIs didn't not require me to head over to the Swagger page since everything was already there, I just didn't notice some private methods
*  I'm getting a `403` response when I try to call the APIs from `/add`, `/edit`, and `reset password`, but I don't think that's a problem :grin:
* I noticed a strange behavior:
 If I click on this observer's `edit` button:
 ```json
{ 
 id:11,
 ngo: "Test NGO",
 phone: "0700000004",
 name: "test4",
}
```

it will redirect me to `/observatori/profil/edit/11`, but what I'm getting for this one is:

```json
{
data: [],
totalItems: 0,
totalPages: 1,
page: 1,
pageSize: 20,
}
```

I think it would be easier to get the observer directly from the store. It may also be more efficient, since it will not make another request to the server.

**Comment by aniri** [comment]
hey @Andrei0872 thanks for the updates! 

Yes, there seems to be a bug when clicking the `edit` button, the endpoint called on the edit screen `GET ​/api​/v1​/observer/?number=PHONE_NO` expects to receive the phone number for that observer and not the id. Could you please update this?

Also, I noticed the add observer phone is missing the password field.

Also, could you please also migrate the import functionality from before? It was on the same page as add observer from before.

Thanks!

**Comment by Andrei0872** [comment]
@aniri ,

> Yes, there seems to be a bug when clicking the edit button, the endpoint called on the edit ...

I think there will be a small problem with this. So, if I reach the `edit` view after clicking the button from the table, then I can grab the observer directly from the store, without making another request. But if I refresh the page while being on the `edit` view, there is no way I can get other information about the observer(not even with a request), apart from its `id`(from the URL). What should I do in such scenario?

> Also, could you please also migrate the import functionality from before?

Where should this go? Should it be another route, or everything in a model?

**Comment by aniri** [comment]
> @aniri ,
> 
> > Yes, there seems to be a bug when clicking the edit button, the endpoint called on the edit ...
> 
> I think there will be a small problem with this. So, if I reach the `edit` view after clicking the button from the table, then I can grab the observer directly from the store, without making another request. But if I refresh the page while being on the `edit` view, there is no way I can get other information about the observer(not even with a request), apart from its `id`(from the URL). What should I do in such scenario?
> 
> > Also, could you please also migrate the import functionality from before?
> 
> Where should this go? Should it be another route, or everything in a model?

@Andrei0872 for the first one, could we use the phone number for now as id? in the url? so it will work? and I'll add a new issue on the backend app to implement a new get by id endpoint :)

For the import could add a new page `/observatori/import` that will open when clicking on the import button? And it should have simple form with id ngo input field, file upload and submit button? It was merged with add observers before, under `/observatori/profil/new`

**Comment by Andrei0872** [comment]
@aniri sure :+1:, thank you

**Comment by Andrei0872** [comment]
Hi @aniri , I've fixed all the _todos_  from above.

File: src/index.html [hunk 1]
```diff
@@ -10,6 +10,7 @@
       content="width=device-width, initial-scale=1, shrink-to-fit=no"
     />
     <link rel="icon" type="image/x-icon" href="favicon.ico" />
+    <link href="https://fonts.googleapis.com/css2?family=Poppins&display=swap" rel="stylesheet">
```
Comment (aniri) [inline src/index.html:13]:
```suggestion
    <link rel="icon" type="image/x-icon" href="favicon.ico" />
```

**Review by aniri — APPROVED** [review]

**Pull request merged**